### PR TITLE
WebGL2 support

### PR DESCRIPTION
**Pull request by astojilj**
This is an experimental feature for now and is disabled by default. It could be enabled by setting MapOptions.isWebGL2.

If MapOptions.isWebGL2 is set, an attempt to create WebGL 2 context happens before falling back to WebGL 1. This way, WebGL 1 legacy browsers are supported (with a warning about unsuccessful WebGL 2 creation) when MapOptions.isWebGL2 is set. 
 
If WebGL2 context is created, `#version 300 es\n` is prepend to glsl shaders, and prelude shaders handle required substitution. e.g. for fragment shaders

```
#if __VERSION__ >= 300
#define varying in
#define gl_FragColor glFragColor
#define texture2D texture
#define textureCube texture
out vec4 glFragColor;
#endif
```

A render test bot running all render tests using WebGL2 is added.

An issue with globe fill extrusion (with terrain disabled) got exposed in the bot: when TERRAIN is defined, centroid is used in vertex shader. However, in that case, we don't call gl.vertexAttrib2f(program.attributes['a_centroid_pos'], 0, 0), and this leads to undefined vertex attrib data (e.g. zero-height-clipping-angle-limit test). Globe 2D shaders don't seem to depend on TERRAIN define, so the original code setting the define is removed.

Screenshot 2023-01-16 at 8 41 51

WebGL1 extensions extTextureHalfFloat and extStandardDerivatives cover functionality that is available by default in WebGL2, so they are modeled as supported, in order to resolve failing heatmap and globe antialiasing tests.

This PR intentionally adds one WebGL2 bot only, to be used for initial validation. Other platforms' bots are to be added to projects using mapbox-gl-js as a submodule.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 ~~- [ ] include before/after visuals or gifs if this PR includes visual changes~~ No visual difference.
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 ~~- [ ] post benchmark scores~~
 - [x] manually test the debug page
 ~~- [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes~~
 ~~- [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port~~
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 ~~- [ ] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`~~
- [x] Document MapOptions.isWebGL2 as experimental

### CODE REVIEW

**Comment by astojilj** [comment]
One render test failure remaining to resolve:

Screenshot 2023-01-16 at 15 21 29


Screenshot 2023-01-16 at 14 39 34

**Comment by endanke** [comment]
Great to see that this little amount of changes were enough to get started! :)

**Review by endanke — APPROVED** [review]
Looks good to me

**Review by karimnaaji — APPROVED** [review]
I have a few suggestions but it looks good to me as well 👍

File: src/render/draw_fill_extrusion.js [hunk 1]
```diff
@@ -70,9 +70,6 @@ function drawExtrusionTiles(painter, source, layer, coords, depthMode, stencilMo
     const baseDefines = ([]: any);
     if (isGlobeProjection) {
         baseDefines.push('PROJECTION_GLOBE_VIEW');
-        if (painter.style.terrainSetForDrapingOnly()) {
-            baseDefines.push('TERRAIN');
-        }
```
Comment (karimnaaji) [inline src/render/draw_fill_extrusion.js:75]: Do you recall the need for this diff?

Comment (astojilj) [reply src/render/draw_fill_extrusion.js:75]: It is about this:

> An issue with globe fill extrusion (with terrain disabled) got exposed in the bot: when TERRAIN is defined, centroid is used in vertex shader. However, in that case, we don't call gl.vertexAttrib2f(program.attributes['a_centroid_pos'], 0, 0), and this leads to undefined vertex attrib data (e.g. zero-height-clipping-angle-limit test). Globe 2D shaders don't seem to depend on TERRAIN define, so the original code setting the define is removed.

defined TERRAIN activates vertex shader path:

https://github.com/mapbox/mapbox-gl-js/blob/67ed1ab99a9ffe9f0f9266ad0ca3ef438d8faf89/src/shaders/fill_extrusion.vertex.glsl#L80-L83

File: src/render/draw_heatmap.js [hunk 2]
```diff
@@ -120,8 +120,9 @@ function bindTextureToFramebuffer(context, painter, texture, fbo, width, height)
     const gl = context.gl;
     // Use the higher precision half-float texture where available (producing much smoother looking heatmaps);
     // Otherwise, fall back to a low precision texture
-    const internalFormat = context.extRenderToTextureHalfFloat ? context.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE;
-    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, width, height, 0, gl.RGBA, internalFormat, null);
+    const type = context.extRenderToTextureHalfFloat ? context.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE;
+    const internalFormat = context.isWebGL2 ? context.extTextureHalfFloat.RGBA16F : gl.RGBA;
+    gl.texImage2D(gl.TEXTURE_2D, 0, internalFormat, width, height, 0, gl.RGBA, type, null);
```
[the remark this answers is not in the document]
Comment (astojilj) [reply src/render/draw_heatmap.js]: > To simplify a bit, could rewrite as:
> 
> to not access it through `extTextureHalfFloat`.

Done. The constant is also acceptable [through gl.context.extRenderToTextureHalfFloat.RGBA16F_EXT](https://registry.khronos.org/webgl/extensions/EXT_color_buffer_half_float/) that would void the need for `/* $FlowFixMe[prop-missing] WebGL2 */`.

File: src/render/draw_heatmap.js [hunk 3]
```diff
@@ -120,8 +120,9 @@ function bindTextureToFramebuffer(context, painter, texture, fbo, width, height)
     const gl = context.gl;
     // Use the higher precision half-float texture where available (producing much smoother looking heatmaps);
     // Otherwise, fall back to a low precision texture
-    const internalFormat = context.extRenderToTextureHalfFloat ? context.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE;
-    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, width, height, 0, gl.RGBA, internalFormat, null);
+    const type = context.extRenderToTextureHalfFloat ? context.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE;
```
[the remark this answers is not in the document]
Comment (astojilj) [reply src/render/draw_heatmap.js]: Both FLOAT and HALF_FLOAT are fine for type, GLES 3.0 and WebGL.

This code is changed to 

`const type = context.extRenderToTextureHalfFloat ? (context.isWebGL2 ? gl.HALF_FLOAT : context.extTextureHalfFloat.HALF_FLOAT_OES) : gl.UNSIGNED_BYTE;`

given that EXT_color_buffer_half_float is WebGL2 extension, too.

File: src/gl/context.js [hunk 4]
```diff
@@ -127,12 +139,12 @@ class Context {
             this.vendor = gl.getParameter(this.extDebugRendererInfo.UNMASKED_VENDOR_WEBGL);
         }
 
-        this.extTextureHalfFloat = gl.getExtension('OES_texture_half_float');
+        this.extTextureHalfFloat = isWebGL2 ? {HALF_FLOAT_OES: 0x140B, RGBA16F: 0x881A} : gl.getExtension('OES_texture_half_float');
```
[the remark this answers is not in the document]
Comment (astojilj) [reply src/gl/context.js]: Rewritten related to issue https://github.com/mapbox/mapbox-gl-js/issues/11327, ensuring also call to gl.getExtension('EXT_color_buffer_half_float'); for WebGL2.

**Pull request merged**